src/sema/expression/mod.rs

```diff
@@ -68,6 +70,17 @@ pub struct ExprContext {
     pub yul_function: bool,
     /// Loops nesting
     pub loops: LoopScopes,
+    /// Stack of currently actively variable scopes
+    pub active_scopes: Vec<VarScope>,
+}
+
+impl ExprContext {
+    pub fn new() -> Self {
+        Self {
+            active_scopes: vec![VarScope(HashMap::new(), None)],
+            ..Default::default()
+        }
+    }
 }
```

Comment:
Having both `fn new()` and `#[derive(Default)]` is confusing. A better solution would be to not `#[derive(Default)]` and do `impl Default for ExprContext`.

Having said all that, how about using the existing `#[derive(Default)]` and do a `enter_scope()` to create the first scope.